# Στοχαστικές Διεργασίες και Βελτιστοποίηση στη Μηχανική Μάθηση

## 7o Εργαστήριο - *Markov Descision Processes*

- Ονομ/νυμο: Χρήστος Νίκου
- AM: 03400146
- Ιδιότητα: Μεταπτυχιακός φοιτητής Επιστήμης Δεδομένων και Μηχανικής Μάθησης (ΕΔΕΜΜ)
- Ηλεκτρονική Διεύθυνση: christosnikou@mail.ntua.gr / chrisnick92@gmail.com

<h1><b>Markov Decision Processes</h1></b>
<p align="justify">Στη συγκεκριμένη άσκηση θα μελετήσετε τους αλγορίθμους <i>Policy Iteration</i> και <i>Value Iteration</i>, καθώς και θα εξοικειωθείτε με βασικές έννοιες των <i>Markov Decision Processes</i>. Οι αλγόριθμοι <i>Policy Iteration</i> και <i>Value Iteration</i> είναι από τους βασικούς αλγορίθμους δυναμικού προγραμματισμού που χρησιμοποιούνται για την επίλυση της εξίσωσης <i>Bellman</i> σε <i>Markov Decision Processes</i>.</p> 
<p align="justify">Το πρόβλημα που θα μελετήσετε είναι αυτό της παγωμένης λίμνης (Frozen Lake) με μέγεθος πλέγματος 8 x 8.</p>


<h2><b>Εξοικείωση με τη βιβλιοθήκη <i>Gym</i></b></h2>

In [ ]:
import numpy as np
import gym
from gym import wrappers

Με την παρακάτω εντολή, ορίζετε το πρόβλημα που θα μελετηθεί:

In [ ]:
env_name = 'FrozenLake8x8-v0'
env = gym.make(env_name)

Με τις παρακάτω εντολές, θα επαναφέρετε τον Agent στην αρχική του θέση και θα οπτικοποιήσετε το πλέγμα και τη θέση του Agent

In [ ]:
env.reset()
env.render()


SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG


Με τις παρακάτω εντολές, ορίζετε την επόμενη ενέργεια με τυχαίο τρόπο και ο Agent κάνει ένα βήμα.

In [ ]:
next_action = env.action_space.sample()
env.step(next_action)
env.render()

  (Left)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG


>*Να εκτελέσετε αρκετές φορές τις τελευταίες εντολές και να παρατηρήσετε κάθε φορά την ενέργεια που ζητείται από τον agent να εκτελέσει και την ενέργεια που αυτός πραγματοποιεί. Πραγματοποιεί πάντοτε ο agent την κίνηση που του ζητείται; Είναι ντετερμινιστικές ή στοχαστικές οι κινήσεις του agent;*

Στο παρακάτω κομμάτι κώδικα επαναφέρουμε τον Agent στην αρχική του θέση και εκτελούμε τις προηγούμενες εντολές όσες φορές δείχνει η μεταβλητή `num_iters`.

In [ ]:
num_iters = 10

env.reset()

for i in range(num_iters):
  print(f"\n{5*'-'}> Move: {i+1} <{5*'-'}\n")
  next_action = env.action_space.sample()
  env.step(next_action)
  env.render()


-----> Move: 1 <-----

  (Down)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 2 <-----

  (Left)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 3 <-----

  (Right)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 4 <-----

  (Down)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 5 <-----

  (Down)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 6 <-----

  (Left)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 7 <-----

  (Down)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 8 <-----

  (Down)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 9 <-----

  (Right)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 10 <-----

  (Right)
SFFFFFFF
FFFFFFFF


Με βάση τις παραπάνω εκτελέσεις παρατηρούμε ότι ο Agent δεν επιλέγει πάντοντε να εκτελέσει την κίνηση που του ζητείται. Για παράδειγμα στην 1η κίνηση του ζητήθηκε να κινηθεί προς το κάτω μέρος του πλέγματος και αυτός παρέμεινε στην ίδια θέση. Το ίδιο παρατηρούμε και στην 3η κίνηση για παράδειγμα που του ζητήθηκε να κινηθεί προς το δεξί μέρος του πλέγματος και αυτός κινήθηκε προς τα κάτω.

Όσον αφορά τη στοχαστικότητα ως προς τις κινήσεις του Agent παρατηρούμε ότι στις κινήσεις 5 και 6 ο Agent βρίσκεται στην ίδια θέση ενώ κατά την 7η κίνηση έχει μετακινηθεί μια θέση δεξιά. Αυτό δείχνει ότι οι κινήσεις του Agent <b>*δεν είναι ντετερμινιστικές*</b> καθώς στη μια περίπτωση παρέμεινε στην ίδια θέση ενώ στην 7η κίνηση μετακινήθηκε προς τα δεξιά. Συνεπώς, παρατηρούμε ότι <b>*υπάρχει στοχαστικότητα*</b> στις κινήσεις του Agent. Η στοχαστικότητα του προβλήματος φαίνεται επίσης αν εκτελέσουμε και πάλι το παραπάνω τμήμα κώδικα και παρατηρήσουμε ότι κατά τη 10η κίνηση ο Agent καταλήγει σε άλλο σημείο του πλέγματος.

In [ ]:
env.reset()

for i in range(num_iters):
  print(f"\n{5*'-'}> Move: {i+1} <{5*'-'}\n")
  next_action = env.action_space.sample()
  env.step(next_action)
  env.render()


-----> Move: 1 <-----

  (Right)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 2 <-----

  (Left)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 3 <-----

  (Left)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 4 <-----

  (Up)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 5 <-----

  (Right)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 6 <-----

  (Up)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 7 <-----

  (Down)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 8 <-----

  (Right)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 9 <-----

  (Up)
SFFFFFFF
FFFFFFFF
FFFHFFFF
FFFFFHFF
FFFHFFFF
FHHFFFHF
FHFFHFHF
FFFHFFFG

-----> Move: 10 <-----

  (Right)
SFFFFFFF
FFFFFFFF
FFFHF

<h2><b>Ερωτήσεις</b></h2>

<ul>
<li>Μελετώντας <a href="https://gym.openai.com/envs/FrozenLake-v0/">αυτό</a> και βασισμένοι στα συμπεράσματα του προηγούμενου ερωτήματος να περιγράψετε σύντομα το πρόβλημα της παγωμένης λίμνης ως πρόβλημα βελτιστοποίησης. Ποιος είναι ο στόχος του agent;</li>
<li>Να διατυπώσετε την ιδιότητα <i>Markov</i>. Πώς απλοποιεί η ιδιότητα αυτή τη μελέτη του συγκεκριμένου προβλήματος;</li>
<li>Να περιγράψετε σύντομα τους αλγορίθμους <i>Policy Iteration</i> και <i>Value Iteration</i>, δίνοντας έμφαση στο διαφορετικό τρόπο με τον οποίο προσεγγίζουν την επίλυση του προβλήματος. Είναι εγγυημένο ότι οι δύο αλγόριθμοι θα συγκλίνουν στη βέλτιστη πολιτική; Αν ναι, οδηγούν σε ίδια ή διαφορετική βέλτιστη πολιτική;</li>
<li>Να εκτελέσετε τα προγράμματα που σας δίνονται, τα οποία επιλύουν το
πρόβλημα της παγωμένης λίμνης, χρησιμοποιώντας τους αλγορίθμους <i>Policy
Iteration</i> και <i>Value Iteration</i> αντίστοιχα. Ποια μέθοδος συγκλίνει στη βέλτιστη λύση σε λιγότερα βήματα; Τι συμπέρασμα βγάζετε; Να ελέγξετε το χρόνο εκτέλεσης του κάθε προγράμματος, χρησιμοποιώντας την εντολή <i>time</i>. Τι συμπέρασμα βγάζετε ως προς την πολυπλοκότητα του κάθε αλγορίθμου;</li>
</ul>

<h2><b>Απαντήσεις</b></h2>

### Ερώτηση 1

>*Μελετώντας <a href="https://www.gymlibrary.ml/environments/toy_text/frozen_lake/">αυτό</a> και βασισμένοι στα συμπεράσματα του προηγούμενου ερωτήματος να περιγράψετε σύντομα το πρόβλημα της παγωμένης λίμνης ως πρόβλημα βελτιστοποίησης. Ποιος είναι ο στόχος του agent;*

Στο πρόβλημα της παγωμένης λίμνης ο Agent προσπαθεί ξεκινώντας από το σημείο του πλέγματος με την ένδειξη Start(S) να φτάσει στο σημείο Goal(G) διασχίζοντας μόνο τα σημεία στα οποία η λίμνη είναι παγωμένη Frozen(F) χωρίς να πέσει στις τρύπες Holes(H) που υπάρχουν στη λίμνη.

- Όπως φαίνεται και παρακάτω ο Agent σε κάθε σημείο του πλέγματος επιλέγει να κάνει μια από τις τέσσερις κινήσεις Up, Down, Right, Left.

In [ ]:
print(f"- Number of possible moves:{env.action_space}")

- Number of possible moves:Discrete(4)


Ο τρόπος με τον οποίο το πρόβλημα της παγωμένης λίμνης εντάσσεται στο πλαίσιο ενός προβλήματος βελτιστοποίσης είναι ότι ο Agent προσπαθεί να εκπαιδευτεί μέσω μιας διαδικασίας ανταμοιβής (reward) για κάθε σωστή κίνηση που εκτελεί με σκοπό να φτάσει στο σημείο της ένδειξης Goal(G). Στο συγκεκριμένο πρόβλημα η αμοιβή που δέχεται ο Agent ανάλογα με τις κινήσεις που εκτελεί φαίνεται παρακάτω.

<b>*Reward schedule:*</b>

- Reach goal(G): +1
- Reach hole(H): 0
- Reach Frozen(F): 0

### Ερώτηση 2

>*Να διατυπώσετε την ιδιότητα Markov. Πώς απλοποιεί η ιδιότητα αυτή τη μελέτη του συγκεκριμένου προβλήματος;*

Σε τελείως φορμαλιστικό επίπεδο η <a href="https://en.wikipedia.org/wiki/Markov_property">*Μαρκοβιανή ιδιότητα*</a> μπορεί συνοψιστεί ως εξής: Αν $(\Omega,\, \mathcal{F},\, \mathbb{P})$ είναι ένα χώρος πιθανότητας, $(S,\mathcal{S})$ ένας μετρήσιμος χώρος και $(\mathcal{F}_i)_{i\in I}$ μια διήθηση$^1$ στο χώρο πιθανότητας $(\Omega,\, \mathcal{F},\, \mathbb{P})$ τότε λέμε ότι η προσαρμοσμένη στη διήθηση  $(\mathcal{F}_i)_{i\in I}$ στοχαστική διαδικασία $X=\{X_t:\Omega\to S\}$ έχει τη *Μαρκοβιανή ιδιότητα* αν για κάθε $A\in \mathcal{S}$ και $s,t\in I$ με $s<t$ ισχύει

$$\mathbb{P}(X_t\in A\,|\,\mathcal{F}_s)=\mathbb{P}(X_t\in A\,|\,X_s).$$

Στη περίπτωση που ο χώρος $S$ είναι διακριτός και $\mathcal{S}$ είναι η διακριτή $\sigma$-άλγεβρα (το δυναμοσύνολο του $S$) τότε η Μαρκοβιανή ιδιότητα παίρνει την εξής γνώριμη μορφή:

$$\mathbb{P}(X_n = x_n\,|\, X_{n-1}=x_{n-1},\dots,X_0=x_0)=\mathbb{P}(X_n=x_n\,|\,X_{n-1}=x_{n-1}).$$


$^1$Με τον όρο διήθηση εννοούμε ότι το σύνολο $I$ είναι ένα <a href="https://en.wikipedia.org/wiki/Total_order">*πλήρως διατεταγμένο*</a> σύνολο δείκτων και κάθε $\mathcal{F}_i$ είναι μια $\sigma-$άλγεβρα στον $(\Omega,\mathbb{P})$ ώστε αν $s,t\in I$ με $s<t$ τότε και $\mathcal{F}_s\subseteq \mathcal{F}_t$.

Χονδρικά, η ιδιότητα Markov μας λέει ότι σε ένα στοχαστικό σύστημα του οποίου οι καταστάσεις περιγράφονται από το σύνολο $S$ η κατάσταση στην οποία θα βρεθεί το σύστημα στο επόμενο βήμα <b>*εξαρτάται μόνο*</b> από το σημείο στο οποίο βρίσκεται το σύστημα και <b>*όχι από τις προηγούμενες καταστάσεις*</b>. Στο πρόβλημα της παγωμένης λίμνης οι καταστάσεις $S$ είναι οι 64 θέσεις του πλέγματος πάνω στις οποίες μπορεί να κινηθεί ο Agent. Το ότι η επόμενη κίνηση που θα κάνει ο Agent εξαρτάται μόνο απ' τη θέση στην οποία βρίσκεται απλοποιεί το πρόβλημα σημαντικά καθώς για να προβλέψουμε το πως θα κινηθεί ο Agent δε χρειάζεται να λάβουμε υπ' όψιν όλο το ιστορικό των κινήσεων που έχει κάνει, πράγμα το οποίο θα ήταν αρκετά πολύπλοκλο ακόμα και στη περίπτωση του $8\times 8$ πλέγματος.

### Ερώτηση 3

>*Να περιγράψετε σύντομα τους αλγορίθμους Policy Iteration και Value Iteration, δίνοντας έμφαση στο διαφορετικό τρόπο με τον οποίο προσεγγίζουν την επίλυση του προβλήματος. Είναι εγγυημένο ότι οι δύο αλγόριθμοι θα συγκλίνουν στη βέλτιστη πολιτική; Αν ναι, οδηγούν σε ίδια ή διαφορετική βέλτιστη πολιτική;*

Πριν περιγράψουμε τους δύο αλγορίθμους *Policy Iteration* και *Value Iteration* που χρησιμοποιούνται στο παρόν Notebook να πούμε δυο λόγια για το ποιο πρόβλημα προσπαθούν να λύσουν.

Οι αλγόριθμοι <b>*Policy Iteration*</b> και <b>*Value Iteration*</b> προσπαθούν να βρουν τη βέλτιστη πολιτική σε ένα πρόβλημα διαδικασιών Μαρκοβιανών αποφάσεων. Πιο συγκεκριμένα, μια Μαρκοβιανή διαδικασία αποφάσεων περιγράφεται από την πεντάδα $(S,A,\{P_{sa}\},\gamma,R)$ όπου:

- $S$ είναι το σύνολο των καταστάσεων (States).

- $A$ είναι το σύνολο των δράσεων (Actions).

- $P_{sa}$ είναι οι πιθανότητες μετάβασης από μια κατάσταση σε κάποια άλλη. Πιο αναλυτικά, για κάθε $s\in S$ και για κάθε $\alpha \in A, P_{s\alpha}$ είναι μια κατανομή πάνω απ' το χώρο καταστάσεων $S$.

- $\gamma \in [0,1)$ είναι ο παράγοντας έκπτωσης (discount factor)

- $R:S\times A\mapsto \mathbb{R}$ είναι η συνάρτηση αμοιβής.

Η δυναμική μιας Μαρκοβιανής διαδικασίας απόφασης είναι η εξής: Αρχικά, το σύστημα ξεκινά από μια αρχική κατάσταση $s_0\in S$ και επιλέγεται μια ενέργεια $\alpha_0\in A$. Ως αποτέλεσμα αυτής της ενέργειας η Μαρκοβιανή διαδικασία επιλέγει κατά τυχαίο τρόμο με βάση το μέτρο πιθανότητας $P_{s_0\alpha_0}$ την επόμενη κατάσταση $s_1$. Σχηματικά, τη διαδικασία αυτή μπορούμε να τη φανταστούμε ως εξής:<br><br>


$$s_0 \overset{\alpha_0}{\mapsto}s_1\overset{\alpha_1}{\mapsto}s_2\overset{\alpha_2}{\mapsto}s_3\overset{\alpha_3}{\mapsto}\dots\,.$$


<br>Έχοντας επισκεφθεί τις καταστάσεις $s_0,s_1,\dots$ με αντίστοιχες ενέργειες $\alpha_0,\alpha_1,\dots,$ το συνολικό κέρδος περιγράφεται μέσω της<br><br>


$$R(s_0,\alpha_0)+\gamma R(s_1,\alpha_1) + \gamma^2R(s_2,\alpha_2)+\dots\, .$$

<br>Ο στόχος της <b>*Eνισχυτικής Μάθησης (Reinforcement Learning)*</b> είναι να μεγιστοποιήσουμε το αναμενόμενο συνολικό κέρδος:<br><br>

$$\tag{3.1}\mathbb{E}\biggl(R(s_0,\alpha_0)+\gamma R(s_1,\alpha_1)+\gamma^2R(s_2,\alpha_2)+\dots \biggr).$$<br><br>

Μια <b>*πολιτική (policy)*</b> είναι μια συνάρτηση $\pi :S\to A$. Λέμε ότι εκτελούμε την πολιτική $\pi$, αν οποτεδήποτε βρισκόμαστε στη κατάσταση $s\in S$, ενεργούμε με βάση την $\alpha = \pi(s)$. Η <b>*συνάρτηση αποτίμησης (value function)*</b> της πολιτικής $\pi$ δίνεται μέσω της<br><br>

$$\tag{3.2}V^{\pi}(s) = \mathbb{E}\biggl(R(s_0,\alpha_0)+\gamma R(s_1,\alpha_1)+\gamma^2 R(s_2,\alpha_2)+\dots\,|\, s_0 = s,\pi\biggr),$$

<br><br> δηλαδή, η $V^{\pi}(s)$ είναι το αναμενόμενο κέρδος ξεκινώντας από την κατάσταση $s$ και λαμβάνοντας αποφάσεις με βάση την πολιτική $\pi$. Δεδομένου μιας πολιτικής $\pi$ η συνάρτηση αμοιβής $V^{\pi}$ ικανοποιεί τις <b>*εξισώσεις Bellman*</b>:<br><br>

$$\tag{3.3}V^{\pi}(s) = R(s) + \gamma \sum_{s'\in S}P_{s\pi(s)}V^{\pi}(s').$$

<br><br> Η <b>*βέλτιση συνάρτηση αποτίμησης (optimal value function)*</b> ορίζεται μέσω της <br><br>

$$\tag{3.4}V^*(s) = \max_{\pi}V^{\pi}(s).$$

<br><br> Η βέλτιση συνάρτηση αποτίμησης ικανοποιεί τις εξής εξισώσεις:<br><br>

$$\tag{3.5}V^*(s) = R(s)+\max_{\alpha \in A}\gamma \sum_{s'\in S}P_{sa}(s')V^*(s').$$

<br><br> Τέλος, ορίζουμε τη <b>*βέλτιστη πολιτική (optimal policy)*</b> $\pi^*:S\to A$ ως εξής<br><br>

$$\tag{3.6} \pi^*(s) = \text{arg max}_{\alpha\in A}\sum_{s'\in S}P_{s\alpha}(s')V^*(s').$$

<br><br> Δηλαδή, η $\pi^*(s)$ μας δίνει σε καθε κατάσταση $s$ την ενέργεια $\alpha \in A$ η οποία μεγιστοποιεί το κέρδος στον δεύτερο προσθετέο της $(3.5)$. Ο σκοπός των αλγορίθμων <b>*Policy iteration*</b> και <b>*Value Iteration*</b> είναι δεδομένου ενός πεπερασμένου συνολου $S$ και πεπερασμένων το πλήθος ενεργειών $\alpha \in A$ να προσδιορίσουν τη βέλτιστη πολιτική $\pi^*$ λύνοντας τις εξισώσεις Bellman της $(3.3)$.

#### *Ο Αλγόριθμος Policy Iteration*

Ο Αλγόριθμος Policy Iteration προσπαθεί να προσδιορίσει τη βέλτιστη πολιτική $\pi^*$ μέσω της εξής επαναληπτικής διαδικασίας που περιγράφεται στον παρακάτω ψευδοκώδικα:

1. Αρχικοποίηση μιας τυχαίας πολιτικής $\pi$
2. Επανάληψη μέχρις ότου ο αλγόριθμος να συγκλίνει:<br>
    (α) $V\leftarrow V^{\pi}.$<br>
    (β) Για κάθε κατάσταση $s$, έστω 
    
$$\pi(s) = \text{arg max}_{\alpha\in A}\sum_{s'\in S}P_{s\alpha}(s')V(s').$$

3. Τέλος επανάληψης
<br>

Έτσι ο αλγόριθμος Policy Iteration υπολογίζει επαναλητικά την συνάρτηση αποτίμησης (value function) και εν συνεχεία ανανεώνει την πολιτική χρησιμοποιώντας αυτή τη συνάρτηση. Στη περίπτωση που $|S|<\infty$ και $|A|<\infty$ η συνάρτηση $V$ θα συγκλίνει σε πεπερασμένα βήματα στη συνάρτηση $V^*$ καθώς και η πολιτική $\pi$ στην $\pi^*$.

Όπως βλέπουμε απ' τα παραπάνω ο αλγόριθμος *Policy Iteration* ξεκινώντας με μια τυχαία πολιτική ανανεώνει σε κάθε επανάληψη την πολιτική παίρνωντας σε κάθε κατάσταση την ενέργεια που μεγιστοποιεί το κέρδος μέσω της συνάρτησης αποτίμησης $V^{\pi}$ που αντιστοιχεί στην πολιτική $\pi$. Παρακάτω συνοψίζουμε και τον αλγόριθμο *Value Iteration*.

#### O Αλγόριθμος Value Iteration

Ο Αλγόριθμος Value Iteration συνοψίζεται στον παρακάτω ψευδοκώδικα.

1. Για κάθε κατάσταση $s\in S$, θέτουμε $V(s)=0$.

2. Επανάληψη μέχρις ότου ο αλγόριθμος να συγκλίνει:

  Για κάθε κατάσταση $s\in S$, ανανέωση $V(s)=R(S)+\max_{\alpha\in A}\gamma \sum_{s'}P_{s\alpha}(s')V(s').$

3. Τέλος επανάληψης

Όπως παρατηρούμε ο αλγόριθμος ανανεώνει σε κάθε επανάληψη την εκτιμώμενη συνάρτηση αποτίμησης χρησιμοποιώντας τις εξισώσεις Bellman της (3.5).

#### Σύγκριση των δύο αλγορίθμων

Συγκρίνοντας τους δύο αλγορίθμους συνοψίζουμε τα εξής:

1. Στον Policy Iteration ο αλγόριθμος περιλαμβάνει την αξιολόγηση της πολιτικής μέσω της συνάρτησης αποτίμησης και βελτιώνει σε κάθε επανάληψη την πολιτική που ακολουθείεται.

2. Σε αντίθεση, ο Value Iteration βρίσκει τη βέλτιστη συνάρτηση αποτίμησης χρησιμοποιώντας επαναληπτικά τις εξισώσεις Bellman (3.5).

3. Και οι δύο αλγόριθμοι συγκλίνουν στη βέλτιστη πολιτική $\pi^*$ σε πεπερασμένα βήματα.

4. Όσον αφορά την πολυπλοκότητα των δύο αλγορίθμων και οι δύο αλγόριθμοι έχουν ίδια πολυπλοκότητα. Παρ'όλα αυτά, στην πράξη για σχετικά μικρές Μαρκοβιανές διαδικασίες απόφασης ο αλγόριθμος Policy Iteration είναι αρκετά γρήγορος και συγκλίνει στη βέλτιστη πολιτική $\pi^*$ σε λίγες επαναλήψεις. Όμως, για Μαρκοβιανές διαδικασίες απόφασης με μεγάλο χώρο καταστάσεων ο ακριβής υπολογισμός της αποτίμησης $V^{\pi}$ είναι αρκετά χρονοβόρος και έτσι σε αυτή την περίπτωση ο Value Iteration μπορεί να είναι καλύτερη επιλογή.

### Ερώτηση 4

>*Να εκτελέσετε τα προγράμματα που σας δίνονται, τα οποία επιλύουν το πρόβλημα της παγωμένης λίμνης, χρησιμοποιώντας τους αλγορίθμους Policy Iteration και Value Iteration αντίστοιχα. Ποια μέθοδος συγκλίνει στη βέλτιστη λύση σε λιγότερα βήματα; Τι συμπέρασμα βγάζετε; Να ελέγξετε το χρόνο εκτέλεσης του κάθε προγράμματος, χρησιμοποιώντας την εντολή time. Τι συμπέρασμα βγάζετε ως προς την πολυπλοκότητα του κάθε αλγορίθμου;*

Παρακάτω εκτελούμε τους δύο αλγορίθμους για το πρόβλημα της παγωμένης λίμνης και συγκρίνουμε τους δύο αλγορίθμους ως προς το χρόνο εκτέλεσης και τα αποτελέσματά τους.

In [ ]:
import time

<h2><b>Policy Iteration</b></h2>

In [ ]:

"""
Solving FrozenLake8x8 environment using Policy iteration.
Author : Moustafa Alzantot (malzantot@ucla.edu)
"""
import numpy as np
import gym
from gym import wrappers


def run_episode(env, policy, gamma = 1.0, render = False):
    """ Runs an episode and return the total reward """
    obs = env.reset()
    total_reward = 0
    step_idx = 0
    while True:
        if render:
            env.render()
        obs, reward, done , _ = env.step(int(policy[obs]))
        total_reward += (gamma ** step_idx * reward)
        step_idx += 1
        if done:
            break
    return total_reward


def evaluate_policy(env, policy, gamma = 1.0, n = 100):
    scores = [run_episode(env, policy, gamma, False) for _ in range(n)]
    return np.mean(scores)

def extract_policy(v, gamma = 1.0):
    """ Extract the policy given a value-function """
    policy = np.zeros(env.nS)
    for s in range(env.nS):
        q_sa = np.zeros(env.nA)
        for a in range(env.nA):
            q_sa[a] = sum([p * (r + gamma * v[s_]) for p, s_, r, _ in  env.P[s][a]])
        policy[s] = np.argmax(q_sa)
    return policy

def compute_policy_v(env, policy, gamma=1.0):
    """ Iteratively evaluate the value-function under policy.
    Alternatively, we could formulate a set of linear equations in iterms of v[s] 
    and solve them to find the value function.
    """
    v = np.zeros(env.nS)
    eps = 1e-10
    while True:
        prev_v = np.copy(v)
        for s in range(env.nS):
            policy_a = policy[s]
            v[s] = sum([p * (r + gamma * prev_v[s_]) for p, s_, r, _ in env.P[s][policy_a]])
        if (np.sum((np.fabs(prev_v - v))) <= eps):
            # value converged
            break
    return v

def policy_iteration(env, gamma = 1.0):
    """ Policy-Iteration algorithm """
    policy = np.random.choice(env.nA, size=(env.nS))  # initialize a random policy
    max_iterations = 200000
    gamma = 1.0
    for i in range(max_iterations):
        old_policy_v = compute_policy_v(env, policy, gamma)
        new_policy = extract_policy(old_policy_v, gamma)
        if (np.all(policy == new_policy)):
            print ('Policy-Iteration converged at step %d.' %(i+1))
            break
        policy = new_policy
    return policy

tic = time.time()
if __name__ == '__main__':
    env_name  = 'FrozenLake8x8-v0'
    env = gym.make(env_name)
    env = env.unwrapped
    optimal_policy = policy_iteration(env, gamma = 1.0)
    scores = evaluate_policy(env, optimal_policy, gamma = 1.0)
    print('Average scores = ', np.mean(scores))
tac = time.time()
print(f"- Execution time for Policy Iteration: {(tac-tic):.6f} sec(s)")

Policy-Iteration converged at step 12.
Average scores =  1.0
- Execution time for Policy Iteration: 3.386103 sec(s)


<h2><b>Value Iteration</b></h2>

In [ ]:
"""
Solving FrozenLake8x8 environment using Value-Itertion.
Author : Moustafa Alzantot (malzantot@ucla.edu)
"""
import numpy as np
import gym
from gym import wrappers


def run_episode(env, policy, gamma = 1.0, render = False):
    """ Evaluates policy by using it to run an episode and finding its
    total reward.
    args:
    env: gym environment.
    policy: the policy to be used.
    gamma: discount factor.
    render: boolean to turn rendering on/off.
    returns:
    total reward: real value of the total reward recieved by agent under policy.
    """
    obs = env.reset()
    total_reward = 0
    step_idx = 0
    while True:
        if render:
            env.render()
        obs, reward, done , _ = env.step(int(policy[obs]))
        total_reward += (gamma ** step_idx * reward)
        step_idx += 1
        if done:
            break
    return total_reward


def evaluate_policy(env, policy, gamma = 1.0,  n = 100):
    """ Evaluates a policy by running it n times.
    returns:
    average total reward
    """
    scores = [
            run_episode(env, policy, gamma = gamma, render = False)
            for _ in range(n)]
    return np.mean(scores)

def extract_policy(v, gamma = 1.0):
    """ Extract the policy given a value-function """
    policy = np.zeros(env.nS)
    for s in range(env.nS):
        q_sa = np.zeros(env.action_space.n)
        for a in range(env.action_space.n):
            for next_sr in env.P[s][a]:
                # next_sr is a tuple of (probability, next state, reward, done)
                p, s_, r, _ = next_sr
                q_sa[a] += (p * (r + gamma * v[s_]))
        policy[s] = np.argmax(q_sa)
    return policy


def value_iteration(env, gamma = 1.0):
    """ Value-iteration algorithm """
    v = np.zeros(env.nS)  # initialize value-function
    max_iterations = 100000
    eps = 1e-20
    for i in range(max_iterations):
        prev_v = np.copy(v)
        for s in range(env.nS):
            q_sa = [sum([p*(r + prev_v[s_]) for p, s_, r, _ in env.P[s][a]]) for a in range(env.nA)] 
            v[s] = max(q_sa)
        if (np.sum(np.fabs(prev_v - v)) <= eps):
            print ('Value-iteration converged at iteration# %d.' %(i+1))
            break
    return v

tic = time.time()
if __name__ == '__main__':
    env_name  = 'FrozenLake8x8-v0'
    gamma = 1.0
    env = gym.make(env_name)
    env = env.unwrapped
    optimal_v = value_iteration(env, gamma);
    policy = extract_policy(optimal_v, gamma)
    policy_score = evaluate_policy(env, policy, gamma, n=1000)
    print('Policy average score = ', policy_score)
tac = time.time()
print(f"- Execution time for Value Iteration: {(tac-tic):.6f} sec(s)")

Value-iteration converged at iteration# 2357.
Policy average score =  1.0
- Execution time for Value Iteration: 2.429732 sec(s)


Στον παρακάτω πίνακα συνοψίζουμε τα αποτελέσματα των δύο αλγορίθμων.

| **Αλγόριθμος** | **Αριθμός Επαναλήψεων** |**Χρόνος Εκτέλεσης (sec)**|
|:---------------|:-----------------|:-----------------|
| **Policy Iteration** |       **12**     |3.386103|
| **Value Iteration**      |       **2357**      |2.429732|

Όπως βλέπουμε και απ' τον παραπάνω πίνακα ο αλγόριθμος Policy Iteration συγκλίνει γρηγορότερα από τον Value Iteration στη βέλτιστη πολιτική. Παρ' όλα αυτα ο αλγόριθμος Value Iteration είναι γρηγορότερος κατά 1 δευτερόλεπτο από τον Policy Iteration.
